### PR TITLE
Hack to allow `bootstrap.sh` to run cleanly on fresh M1

### LINKED ISSUE #9722: Bug Report: Cannot run bootstrap.sh on arm64 mac
### Overview of the Issue

When running `./bootstrap.sh` fresh on an M1 mac, I get the following error:

```
➜  vitess git:(vmain) rm -rf dist/vt-protoc-*
➜  vitess git:(vmain) make minimaltools 
Thu Feb 17 06:12:44 EST 2022: Installing minimal dependencies
installing protoc 3.19.4
ERROR: unsupported architecture
make: *** [minimaltools] Error 1
```

The issue is with the platform/arch checking done within `install_protoc` [here](https://github.com/vitessio/vitess/blob/ae72a9b03a39dab563de2e735e85d99dcb5951c0/bootstrap.sh#L100-L109)

### Reproduction Steps

```markdown
1. remove any previous bootstrap artifacts from `./dist`
2. run `./bootstrap.sh` and/or `make minimaltools`
```

### Binary Version

```shell
On current `main`
```

### Operating System and Environment details

```shell
➜  vitess git:(vmain) uname -sm      
Darwin arm64
```
```


### Log Fragments

_No response_

### PR DESCRIPTION
**Pull request by ajm188**
## Description

Currently, we don't provide a `protoc` binary for osx/arm64
combinations (i.e. "M1 Macs"), and neither does upstream
protobuf.

The current upstream suggested workaround is to build from source or use
unsigned binaries.

Luckily, if you have Rosetta installed on your machine (which for a while
was required to [run `man` without errors](https://apple.stackexchange.com/questions/430310/apple-silicon-usr-bin-mandoc-bad-cpu-type-in-executable/430311#430311), it's actually totally fine to run the x86-64 version.

So, we just warn the user we're doing this and install the other binary.

### Testing

```
➜  vitess git:(andrew/protoc-m1) ✗ rm -rf dist/vt-protoc-*    
➜  vitess git:(andrew/protoc-m1) ✗ 
➜  vitess git:(andrew/protoc-m1) ✗ make minimaltools >/dev/null
WARN: Installing x86_64 protoc on arm macos; This will work if you have Rosetta installed. See https://github.com/protocolbuffers/protobuf/issues/9397.
--2022-02-17 06:08:00--  https://github.com/vitessio/vitess-resources/releases/download/v1.0/protoc-3.19.4-osx-x86_64.zip
Resolving github.com (github.com)...
[wget output elided]
Saving to: ‘protoc-3.19.4-osx-x86_64.zip’

protoc-3.19.4-osx-x86_64.zip       100%[================================================================>]   2.55M  12.0MB/s    in 0.2s    

2022-02-17 06:08:01 (12.0 MB/s) - ‘protoc-3.19.4-osx-x86_64.zip’ saved [2676128/2676128]
```

## Related Issue(s)

Fixes #9722

## Checklist
- [x] Should this PR be backported? **no**
- [x] Tests were added or are not required **n/a**
- [x] Documentation was added or is not required **n/a**

### CODE REVIEW

**Review by deepthi — COMMENTED** [review]
I know you can `brew install` the binaries for `arm64`. Are those unsigned?

**Comment by deepthi** [comment]
> So this is for `protoc`, _not_ for `protoc-gen-go`. I don't see `protoc` in homebrew:
> 
> ```
> ➜  local git:(andrew/flag-deprecation) ✗ brew search protoc
> ==> Formulae
> libbitcoin-protocol       protoc-gen-go-grpc        protoc-gen-grpc-web       cproto
> libsignal-protocol-c      protoc-gen-gogo           spice-protocol
> protoc-gen-go             protoc-gen-gogofaster     wayland-protocols
> ```

Installing `protoc-gen-go` also gets you `protoc`.

**Comment by derekperkins** [comment]
Can we just switch to buf?

**Comment by ajm188** [comment]
> Installing `protoc-gen-go` also gets you `protoc`.

Ohhh TIL!! Cool let's go with homebrew then.

**Comment by ajm188** [comment]
Here's the testing of the homebrew version of the installation. Ran without issue on my machine.

```
➜  vitess git:(andrew/protoc-m1) ✗ brew uninstall protobuf && rm -rf dist/vt-protoc-3.19.4
Uninstalling /opt/homebrew/Cellar/protobuf/3.19.4... (270 files, 19.8MB)
➜  vitess git:(andrew/protoc-m1) ✗ ./bootstrap.sh
installing protoc 3.19.4
WARN: Protobuf does not have a protoc binary for arm64 macos.
Checking for homebrew installation. Altertatively, you may install the x86-64
version if you have Rosetta installed; your mileage may vary.

See https://github.com/protocolbuffers/protobuf/issues/9397.
==> Downloading https://ghcr.io/v2/homebrew/core/protobuf/manifests/3.19.4
Already downloaded: /Users/andrew/Library/Caches/Homebrew/downloads/71012f077ca8dd7359a17c5aad9174eec4489a851f49a8f36109765610cfbe43--protobuf-3.19.4.bottle_manifest.json
==> Downloading https://ghcr.io/v2/homebrew/core/protobuf/blobs/sha256:a467da7231471d7
Already downloaded: /Users/andrew/Library/Caches/Homebrew/downloads/2b07e1c4f6601bac9d9b643dd1f493acf3710606b3db8feec3e3d5f056af040a--protobuf--3.19.4.arm64_monterey.bottle.tar.gz
==> Pouring protobuf--3.19.4.arm64_monterey.bottle.tar.gz
==> Caveats
Emacs Lisp files have been installed to:
  /opt/homebrew/share/emacs/site-lisp/protobuf
==> Summary
🍺  /opt/homebrew/Cellar/protobuf/3.19.4: 270 files, 19.8MB
==> Running `brew cleanup protobuf`...
Disable this behaviour by setting HOMEBREW_NO_INSTALL_CLEANUP.
Hide these hints with HOMEBREW_NO_ENV_HINTS (see `man brew`).
➜  vitess git:(andrew/protoc-m1) ✗ which protoc
/opt/homebrew/bin/protoc
➜  vitess git:(andrew/protoc-m1) ✗ ls -l bin/protoc
lrwxr-xr-x  1 andrew  staff  47 Feb 18 09:27 bin/protoc -> /opt/homebrew/Cellar/protobuf/3.19.4/bin/protoc
➜  vitess git:(andrew/protoc-m1) ✗ ls -l $(which protoc)
lrwxr-xr-x  1 andrew  admin  36 Feb 18 09:27 /opt/homebrew/bin/protoc -> ../Cellar/protobuf/3.19.4/bin/protoc
➜  vitess git:(andrew/protoc-m1) ✗ ls dist/vt-protoc-3.19.4 
bin     include
➜  vitess git:(andrew/protoc-m1) ✗ make proto
Fri Feb 18 09:48:47 EST 2022: Installing minimal dependencies
skipping protoc install. remove /Users/andrew/dev/vitess/dist/vt-protoc-3.19.4 to force re-install.
/opt/homebrew/bin/etcd
etcd already installed
skipping k3s install. remove /Users/andrew/dev/vitess/dist/k3s to force re-install.

bootstrap finished - run 'make build' to compile
Fri Feb 18 09:49:04 EST 2022: Compiling proto definitions
```

File: bootstrap.sh [hunk 1]
```diff
@@ -105,13 +105,44 @@ install_protoc() {
...
+
+See https://github.com/protocolbuffers/protobuf/issues/9397.
+WARNING
+
+    if [[ -z "$(command -v brew)" ]]; then
+      echo "Could not find \`brew\` command. Please install homebrew and retry." >&2;
+      exit 1;
+    fi
+
+    brew install protobuf;
```
Comment (deepthi) [inline bootstrap.sh:130]: For linux we always install a well-known version (`$version`). Should we do the same for OSX and pin a known version?

Comment (ajm188) [reply bootstrap.sh:130]: you actually ... can't 🙈 

with homebrew you just get latest, which is one of the bigger downsides of going this way vs the "just download the x86 and ask them to use rosetta" that i originally advocated

e.g.

```
➜  dev brew list protobuf | head -n 1
/opt/homebrew/Cellar/protobuf/3.19.4/INSTALL_RECEIPT.json
➜  dev brew search protobuf
==> Formulae
protobuf ✔          protobuf-c          protobuf@3.6        swift-protobuf
➜  dev brew uninstall protobuf >/dev/null
➜  dev brew install protobuf@3.19.4
Warning: No available formula with the name "protobuf@3.19.4". Did you mean protobuf@3.6?
==> Searching for similarly named formulae...
This similarly named formula was found:
protobuf@3.6
To install it, run:
  brew install protobuf@3.6
==> Searching for a previously deleted formula (in the last month)...
Error: No previously deleted formula found.
==> Searching taps on GitHub...
Error: No formulae found in taps.
➜  dev
```

**Review by deepthi — APPROVED** [review]
I'm torn between the two options (brew and rosetta), they both have upsides and downsides.
I'll let you choose, just LMK once this is final so that we can merge it.

**Comment by ajm188** [comment]
> I'm torn between the two options (brew and rosetta), they both have upsides and downsides.

Agreed. I'm going to advocate for the Homebrew version because it feels (with nothing to actually back this up) slightly less sketchy/hacky to me, plus:
1. there are probably no serious differences between code generated between two patch versions (#9644 should actually mitigate this concern completely for our purposes) AND
1. we shouldn't be in this situation (no arm binaries) for too much longer.

I'll get the tests passing and then merge (assuming they're flakes; if I need to make code changes I'll ping you again)

**Comment by dbussink** [comment]
> 1. there are probably no serious differences between code generated between two patch versions ([Check for proto drift in the CI #9644](https://github.com/vitessio/vitess/pull/9644) should actually mitigate this concern completely for our purposes) AND

Turns out this is the case in the sense that the version number is in generated code, so any time someone locally updates the protobuf and regenerates code, it would be considered "broken" as the Homebrew version diverges and is different and that generates a diff.

There's however now a binary out with also M1 support, so we can update this in https://github.com/vitessio/vitess/pull/10803 :smile:.

**Comment by ajm188** [comment]
> Turns out this is the case

:lolsob: ah the hubris of february. thanks for the fix!!

**Pull request merged**